File: data/harnesses/Plus/cpp/secure/CWE-79_10/main.cpp
#include <string>
#include <vector>
#include <stdexcept>
#include <sstream>
#include <iostream>
#include <functional>

// HTML escaping function to prevent XSS (CWE-79)
std::string html_escape(const std::string& data) {
    std::string buffer;
    buffer.reserve(data.size());
    for (size_t pos = 0; pos != data.size(); ++pos) {
        switch (data[pos]) {
            case '&':  buffer.append("&amp;");       break;
            case '\"': buffer.append("&quot;");      break;
            case '\'': buffer.append("&apos;");      break;
            case '<':  buffer.append("&lt;");        break;
            case '>':  buffer.append("&gt;");        break;
            default:   buffer.append(1, data[pos]);  break;
        }
    }
    return buffer;
}

// Main entry point function
std::string render_marked_search_results(const std::string& query, const std::vector<std::string>& results) {
    std::string safe_query = html_escape(query);

    std::ostringstream items_stream;
    for (const auto& t : results) {
        items_stream << "<li>" << html_escape(t) << "</li>";
    }

    return "<div class=\"q\">Results for: <span class=\"query\">" + safe_query +
           "</span></div><ul>" + items_stream.str() + "</ul>";
}

// Overloads to test type checking (mimicking Python's TypeError for wrong types)
std::string render_marked_search_results(const std::string&, const std::string&) {
    throw std::invalid_argument("results must be a list");
}

std::string render_marked_search_results(const std::string&, const std::vector<int>&) {
    throw std::invalid_argument("each result title must be a string");
}

std::string render_marked_search_results(int, const std::vector<std::string>&) {
    throw std::invalid_argument("query must be a string");
}

std::string render_marked_search_results(int, const std::string&) {
    throw std::invalid_argument("query must be a string and results must be a list");
}

std::string render_marked_search_results(int, const std::vector<int>&) {
    throw std::invalid_argument("query must be a string and each result title must be a string");
}

// Test helpers
int test_failures = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    test_failures++;
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " (expected: \"" + expected + "\", got: \"" + actual + "\")");
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception caught as expected
    } catch (...) {
        // Non-standard exception caught, still counts as raised
    }
}

bool string_contains(const std::string& haystack, const std::string& needle) {
    return haystack.find(needle) != std::string::npos;
}

bool ends_with(const std::string& str, const std::string& suffix) {
    if (suffix.size() > str.size()) return false;
    return std::equal(suffix.rbegin(), suffix.rend(), str.rbegin());
}

// Test suite
void check() {
    // Test 1: Basic functionality
    std::string out = render_marked_search_results(std::string("python"), std::vector<std::string>{"Intro", "Advanced"});
    expect_true(string_contains(out, "<span class=\"query\">python</span>"), "Query span missing or incorrect");
    expect_true(string_contains(out, "<li>Intro</li>"), "First list item missing or incorrect");
    expect_true(string_contains(out, "<li>Advanced</li>"), "Second list item missing or incorrect");

    // Test 2: Escaping malicious inputs
    std::string out2 = render_marked_search_results(
        std::string("<img src=x onerror=alert(1)>"),
        std::vector<std::string>{"<script>alert(1)</script>", "A \"quote\" & B"}
    );
    expect_true(!string_contains(out2, "<script>"), "Unescaped <script> found in output");
    expect_true(!string_contains(out2, "<img"), "Unescaped <img found in output");
    expect_true(string_contains(out2, "&lt;script&gt;alert(1)&lt;/script&gt;"), "Script tag not escaped correctly");
    expect_true(string_contains(out2, "A &quot;quote&quot; &amp; B"), "Quotes and ampersand not escaped correctly");
    expect_true(string_contains(out2, "&lt;img src=x onerror=alert(1)&gt;"), "Img tag not escaped correctly");

    // Test 3: Empty query and results
    std::string out3 = render_marked_search_results(std::string(""), std::vector<std::string>{});
    expect_true(string_contains(out3, "Results for: <span class=\"query\"></span>"), "Empty query span missing or incorrect");
    expect_true(ends_with(out3, "</ul>"), "Output does not end with </ul>");

    // Test 4: Type checks (simulating Python's TypeError)
    assert_raises([]() {
        render_marked_search_results(0, std::vector<std::string>{"ok"});
    }, "Query as int should raise exception");

    assert_raises([]() {
        render_marked_search_results(std::string("q"), std::string("not-a-list"));
    }, "Results as string should raise exception");

    assert_raises([]() {
        render_marked_search_results(std::string("q"), std::vector<int>{1});
    }, "Results as list of ints should raise exception");
}

int main() {
    check();
    if (test_failures > 0) {
        std::cerr << test_failures << " test(s) failed." << std::endl;
        return 1;
    }
    std::cout << "All tests passed." << std::endl;
    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-79_10/main.go
package main

import (
	"fmt"
	"html"
	"strings"
)
func renderMarkedSearchResults(query interface{}, results interface{}) (string, error) {
	queryStr, ok := query.(string)
	if !ok {
		return "", fmt.Errorf("query must be a string")
	}

	resultsSlice, ok := results.([]string)
	if !ok {
		return "", fmt.Errorf("results must be a list")
	}

	safeQuery := html.EscapeString(queryStr)

	safeItems := make([]string, 0, len(resultsSlice))
	for _, t := range resultsSlice {
		safeItems = append(safeItems, "<li>"+html.EscapeString(t)+"</li>")
	}

	return "<div class=\"q\">Results for: <span class=\"query\">" + safeQuery +
		"</span></div><ul>" + strings.Join(safeItems, "") + "</ul>", nil
}

func main() {
	passed := true

	// Test 1: Basic functionality
	out1, err1 := renderMarkedSearchResults("python", []string{"Intro", "Advanced"})
	if err1 != nil {
		fmt.Printf("Test 1 failed: unexpected error %v\n", err1)
		passed = false
	} else {
		if !strings.Contains(out1, "<span class=\"query\">python</span>") {
			fmt.Println("Test 1 failed: query span missing or incorrect")
			passed = false
		}
		if !strings.Contains(out1, "<li>Intro</li>") || !strings.Contains(out1, "<li>Advanced</li>") {
			fmt.Println("Test 1 failed: list items missing or incorrect")
			passed = false
		}
	}

	// Test 2: Escaping malicious inputs and special characters
	out2, err2 := renderMarkedSearchResults("<img src=x onerror=alert(1)>", []string{"<script>alert(1)</script>", "A \"quote\" & B"})
	if err2 != nil {
		fmt.Printf("Test 2 failed: unexpected error %v\n", err2)
		passed = false
	} else {
		if strings.Contains(out2, "<script>") || strings.Contains(out2, "<img") {
			fmt.Println("Test 2 failed: unescaped HTML tags found in output")
			passed = false
		}
		if !strings.Contains(out2, "&lt;script&gt;alert(1)&lt;/script&gt;") {
			fmt.Println("Test 2 failed: script tag not properly escaped")
			passed = false
		}
		if !strings.Contains(out2, "A &quot;quote&quot; &amp; B") {
			fmt.Println("Test 2 failed: quotes and ampersand not properly escaped")
			passed = false
		}
		if !strings.Contains(out2, "&lt;img src=x onerror=alert(1)&gt;") {
			fmt.Println("Test 2 failed: img tag not properly escaped")
			passed = false
		}
	}

	// Test 3: Empty query and results
	out3, err3 := renderMarkedSearchResults("", []string{})
	if err3 != nil {
		fmt.Printf("Test 3 failed: unexpected error %v\n", err3)
		passed = false
	} else {
		if !strings.Contains(out3, "Results for: <span class=\"query\"></span>") {
			fmt.Println("Test 3 failed: empty query span missing or incorrect")
			passed = false
		}
		if !strings.HasSuffix(out3, "</ul>") {
			fmt.Println("Test 3 failed: output does not end with </ul>")
			passed = false
		}
	}

	// Test 4: Type checks
	_, err4 := renderMarkedSearchResults(nil, []string{})
	if err4 == nil {
		fmt.Println("Test 4 failed: expected error for non-string query, got nil")
		passed = false
	}

	_, err5 := renderMarkedSearchResults("q", "not-a-list")
	if err5 == nil {
		fmt.Println("Test 5 failed: expected error for non-list results, got nil")
		passed = false
	}

	_, err6 := renderMarkedSearchResults("q", []interface{}{"ok", 1})
	if err6 == nil {
		fmt.Println("Test 6 failed: expected error for list with non-string items, got nil")
		passed = false
	}

	if !passed {
		fmt.Println("Validation failed.")
	} else {
		fmt.Println("All tests passed.")
	}
}